### PR TITLE
Add async methods for auth-lib

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java [hunk 1]
```diff
@@ -5,66 +5,82 @@
...
-        return isAvailable();
-    }
+    protected abstract TokenCredential createTokenCredential();
 
     public AzureEnvironment getEnvironment() {
         return entity == null ? null : entity.getEnvironment();
     }
 
+    boolean checkAvailable() {
+        if (this.entity.isAvailable()) {
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:71]: Why we refresh the available status only if it is not available? Will it always be true once available?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java [hunk 2]
```diff
@@ -95,7 +111,65 @@ public boolean isAvailable() {
...
+            return this;
+        }).flatMap(ignore -> {
+            if (this.entity.getSubscriptions() == null) {
+                return initializeSubscriptions().then(Mono.just(this));
+            }
+            return Mono.just(this);
+        });
+    }
+
+    private String getEmailAndRefreshToken(AccessToken accessToken) {
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:142]: Shall we split the method into `refershEmail` and `getRefreshToken`? As now it do return the refresh token, but modify the email address with side effect.

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java [hunk 3]
```diff
@@ -95,7 +111,65 @@ public boolean isAvailable() {
...
+    private Mono<Tuple2<List<String>, AccessToken>> verifyAndGetTenantId(TokenCredential credential) {
+        AzureEnvironment env = ObjectUtils.firstNonNull(this.entity.getEnvironment(), Azure.az(AzureCloud.class).getOrDefault());
+        TokenRequestContext tokenRequestContext = new TokenRequestContext()
+                .addScopes(ScopeUtil.resourceToScopes(env.getManagementEndpoint()));
+        Mono<AccessToken> mono = credential instanceof TenantCredential ?
+                ((TenantCredential) credential).getAccessToken(null, tokenRequestContext) :
+                credential.getToken(tokenRequestContext);
+
+        return mono.flatMap(accessToken ->
+                Mono.zip(AzureResourceManager.authenticate(request -> Mono.just(accessToken), new AzureProfile(env))
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:171]: Please refer https://github.com/microsoft/azure-maven-plugins/blob/feature-auth/azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AzureAppService.java#L135 to set the user agent with track 2 SDK, we may need to do this in the common library

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java (hunk 1 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:71]: for avail: one check and then remain unchanged

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java (hunk 2 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:142]: there has shared code 
```
        IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
        if (result == null) {
            return null;
        }

```

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java [hunk 4]
```diff
@@ -46,129 +49,140 @@ public Account account() throws AzureToolkitAuthenticationException {
...
 
-    public AzureAccount login(AuthType type) throws LoginFailureException {
-        AuthConfiguration auth = new AuthConfiguration();
-        auth.setType(type);
-        loginWithAuthConfiguration(auth);
-        return this;
+    public Mono<Account> authenticateAccounts(List<Account> accounts) {
+        Mono<Account> current = accounts.get(0).authenticateAsync();
+        for (int i = 1; i < accounts.size(); i++) {
+            current = current.onErrorResume(e -> accounts.get(0).authenticateAsync());
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java:82]: -> `e -> accounts.get(i).authenticateAsync()` ?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java (hunk 3 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:171]: user agent should be in global config, let's add a todo comment here(just keep this PR not far away from auth)

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java (hunk 4 above)
Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java:82]: good catch

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java (hunk 2 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:142]: Yes, but it is a bit confused that one get method for both email and refresh token

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java [hunk 5]
```diff
@@ -95,7 +111,65 @@ public boolean isAvailable() {
...
+            refreshTokenFromResult = (String) FieldUtils.readField(result, "refreshToken", true);
+        } catch (IllegalAccessException e) {
+            throw new AzureToolkitAuthenticationException("Cannot read refreshToken from IAuthenticationResult.");
+        }
+
+        return refreshTokenFromResult;
+    }
+
+    private Mono<Tuple2<List<String>, AccessToken>> verifyAndGetTenantId(TokenCredential credential) {
+        AzureEnvironment env = ObjectUtils.firstNonNull(this.entity.getEnvironment(), Azure.az(AzureCloud.class).getOrDefault());
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:163]: Do we need to handle the issue that entity environment differs from `AzureAccount` value?

Comment (andxu) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:163]: in some cases, the env is from auth method, like azure cli, vscode

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java [hunk 6]
```diff
@@ -112,7 +186,7 @@ public boolean isAuthenticated() {
     public void selectSubscription(List<String> selectedSubscriptionIds) {
         requireAuthenticated();
         if (CollectionUtils.isEmpty(selectedSubscriptionIds)) {
-            throw new IllegalArgumentException("Cannot select subscription by an empty id list.");
+            throw new IllegalArgumentException("Cannot select subscription by empty subscription id list.");
```
Comment (Flanker32) [inline azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:189]: `You must select at least one subscription`?

File: azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java (hunk 5 above)
Comment (Flanker32) [reply azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java:163]: In that case, should we change the account env value after authenticated?

**Pull request closed without merging**